### PR TITLE
Allow onClick on next/link component's child

### LINKED ISSUE #1490: Consider handling onClick on next/link elements
Currently `next/link` will ignore `onClick` props set on `<Link>` or contained `<a>` elements:

``` js
import Link from 'next/link'

const customHandleClick = () => console.log('Free pizza!')

// Somewhere in render-land

<Link href='/' onClick={customHandleClick}>
  <a>No free pizza :(</a>
</Link>

<Link href='/'>
  <a onClick={customHandleClick}>No free pizza :/</a>
</Link>
```

Clicking either of those two links will not call the `customHandleClick` function.

This is because the `onClick` handler of the contained `<a>` tag is overridden by the `next/link` implementation [here](https://github.com/zeit/next.js/blob/32af8294a7c2b3a39c8b04c6d751fb8a46875077/lib/link.js#L116).

If this were to be changed a good way would be checking for, and calling, the `onClick` prop passed to the `<Link>` element as part of its internal click handler. See [react-router’s Link element](https://github.com/ReactTraining/react-router/blob/73feda30b59731bc89a0a419a6be92dd93b6f0a5/packages/react-router-dom/modules/Link.js#L35-L44) for inspiration.

Why I would consider changing this: `onClick` handlers on links can be useful for various things, but mostly I didn’t expect that the `onClick` wasn’t being called at all. I’ve added support for this to my custom `Link` element, but I think `onClick` "should just work".

### PR DESCRIPTION
**Pull request by typeofweb**
Allow `onClick` on `next/link` child. This should not be a breaking change, but it's a very useful feature. Real-life use cases include: analytics or closing menu on navigation, and other.

- [x] allow optional `onClick` on `next/link` component's child
- [x] call original `child.props.onClick(e)` before `this.linkClicked(e)`
- [x] add integration tests
- [x] cancel the navigation if `e.defaultPrevented === true`

Fixes #1490

### CODE REVIEW

File: lib/link.js [hunk 1]
```diff
@@ -127,7 +127,12 @@ export default class Link extends Component {
     // This will return the first child, if multiple are provided it will throw an error
     const child = Children.only(children)
     const props = {
-      onClick: this.linkClicked
+      onClick: (e) => {
+        if (child.props && typeof child.props.onClick === 'function') {
```
Comment (timneutkens) [inline lib/link.js]: Smart 👍

**Review by timneutkens — CHANGES_REQUESTED** [review]
Besides the tests failing and preventDefault not being handled this PR looks great to me 👍

**Comment by mmiszy** [comment]
Yes, about those tests… I'm not sure why they failed ;) I'll dig into that tomorrow.
Thanks for the review!

**Comment by timneutkens** [comment]
If you want to run a single test you can do it like this:

```
yarn testonly --testPathPattern "basic" -t "delete a page and add it back"
```

Where `basic` is the test suite (integration/basic) and `delete a page and add it back` is the specific test you want to run.

**Comment by brandonjcreek** [comment]
Keeping my eye on this.  This is one of those minor pain points of Next that creep up on me more often times than not.

**Review by timneutkens — APPROVED** [review]
💯 awesome 👌

**Comment by timneutkens** [comment]
Thank you very much!

**Comment by mmiszy** [comment]
Thank YOU 👍

**Pull request merged**